Moove.AI Layer Feature

drewjj opened the pull request:
<!--- Thanks for the contribution! -->

# PR Details

## Description

<!--- Describe your changes in detail -->
This includes the Moove.AI layer support for the CV Manager. 

- Updates to the webapp have been made: new slice and new layer on the map page.
- New API endpoint for moove.ai has been added.
- New environment variables.
- Moove.ai feature flag to turn the feature on or off.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been tested locally with a GCP service account key that has permission to query the Moove.AI BigQuery database.

Using the following new environment variables:
ENABLE_MOOVE_AI_FEATURES='true'
GOOGLE_ACCESS_KEY_NAME=(secret)
GCP_PROJECT_ID=(secret)
MOOVE_AI_SEGMENT_AGG_STATS_TABLE=(secret)
MOOVE_AI_SEGMENT_EVENT_STATS_TABLE=(secret)

![image](https://github.com/user-attachments/assets/202823d1-7710-4c95-ac5d-1ce73fe6544e)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If a section applies, ensure you have met all of the associated checks -->

- [x] My changes require new environment variables:
  - [x] I have updated the docker-compose, K8s YAML, and all dependent deployment configuration files.
- [x] My changes require updates to the documentation:
  - [x] I have updated the documentation accordingly.
- [x] My changes require updates and/or additions to the unit tests:
  - [x] I have modified/added tests to cover my changes.
- [x] All existing tests pass.

drewjj committed:
Moove AI endpoint added to the API

drewjj committed:
Webapp updates for displaying moove ai data

drewjj committed:
Moove AI webapp tests

drewjj committed:
Add API unit tests for moove ai endpoint

drewjj committed:
Update docker compose files

drewjj committed:
Merge remote-tracking branch 'origin/develop' into moove-ai-support

drewjj committed:
Include new env vars in README

jacob6838 commented on webapp/src/apis/rsu-api.ts:
```diff
@@ -129,6 +130,15 @@ class RsuApi {
     })
 
+  // Moove AI
+  postMoveAiData = async (token: string, body: Object, url_ext: string = ''): Promise<ApiMsgRespWithCodes<any>> =>
```
suggestion (non-blocking) - Add type def for the returned data

drewjj replied:
Addressed

jacob6838 commented on webapp/src/pages/Map.tsx:
```diff
@@ -239,6 +256,28 @@ function MapPage() {
 
   const [expandedLayers, setExpandedLayers] = useState<string[]>([])
 
+  // Moove AI layer local state variables
+  const [mooveAiPolygonSource, setMooveAiPolygonSource] = useState<GeoJSON.Feature<GeoJSON.Geometry>>({
+    type: 'Feature',
+    geometry: {
+      type: 'Polygon',
+      coordinates: [],
+    },
+    properties: {},
+  })
+
+  const [mooveAiPolygonPointSource, setMooveAiPolygonPointSource] = useState<
+    GeoJSON.FeatureCollection<GeoJSON.Geometry>
+  >({
+    type: 'FeatureCollection',
+    features: [],
+  })
+
+  const [mooveAiDataSource, setMooveAiDataSource] = useState<GeoJSON.FeatureCollection<GeoJSON.Geometry>>({
+    type: 'FeatureCollection',
+    features: [],
+  })
```
suggestion (non-blocking) - Move this logic to the mapSlice, this file is massive...

drewjj replied:
Addressed

jacob6838 commented on webapp/src/pages/Map.tsx:
```diff
@@ -354,10 +393,59 @@ function MapPage() {
+
+    setMooveAiPolygonPointSource((prevPointSource) => ({
+      ...prevPointSource,
+      features: pointSourceFeatures,
+    }))
+
+    // Get coordinates including preview point if it exists
+    let polygonCoords = [...mooveAiCoordinates]
+    if (previewPoint && addMooveAiPoint) {
+      const previewCoords = previewPoint.geometry.coordinates
+
+      if (polygonCoords.length >= 3 && polygonCoords[0] === polygonCoords[polygonCoords.length - 1]) {
+        // For completed polygon: Remove closing point, add preview, then close
+        polygonCoords = polygonCoords.slice(0, -1)
+        polygonCoords.push(previewCoords)
+        polygonCoords.push(polygonCoords[0])
+      } else if (polygonCoords.length === 2) {
+        // For two points: Draw triangle with preview point
+        polygonCoords.push(previewCoords)
+        polygonCoords.push(polygonCoords[0])
+      } else if (polygonCoords.length === 1) {
+        // For one point: Draw line to preview point
+        polygonCoords = [[...polygonCoords[0]], [...previewCoords]] // Create a fresh array with both points
+      }
+    } else if (polygonCoords.length >= 3) {
+      // Close the polygon if we have 3+ points and no preview
+      polygonCoords.push(polygonCoords[0])
+    }
+
+    setMooveAiPolygonSource(
+      (prevPolygonSource) =>
+        ({
+          ...prevPolygonSource,
+          geometry: {
+            type: polygonCoords.length === 2 ? 'LineString' : 'Polygon', // Use LineString for 2 points
+            coordinates: polygonCoords.length === 2 ? polygonCoords : [polygonCoords],
+          },
+        } as GeoJSON.Feature<GeoJSON.Geometry>)
+    )
+  }, [mooveAiCoordinates, activeLayers, addMooveAiPoint, previewPoint])
```
suggestion (non-blocking) - refactor this to use useMemo -> it looks like mooveAiPolygonPointSource and mooveAiPolygonSource are only updated here, so no need to re-render unnecessarily with useEffect

drewjj replied:
Addressed

jacob6838 commented on webapp/src/pages/Map.tsx:
```diff
@@ -354,10 +393,59 @@ function MapPage() {
+  // Effect for handling polygon updates moove-ai-layer
+  useEffect(() => {
+    if (!activeLayers.includes('moove-ai-layer')) return
+    const pointSourceFeatures: Array<GeoJSON.Feature<GeoJSON.Geometry>> = []
+
+    mooveAiCoordinates.forEach((point) => {
+      pointSourceFeatures.push(createPointFeature(point))
+    })
+
+    setMooveAiPolygonPointSource((prevPointSource) => ({
+      ...prevPointSource,
+      features: pointSourceFeatures,
+    }))
```
Simplify
```suggestion
    setMooveAiPolygonPointSource((prevPointSource) => ({
      ...prevPointSource,
      features: mooveAiCoordinates.map(createPointFeature),
    }))
```

drewjj replied:
Addressed

jacob6838 commented on webapp/src/pages/Map.tsx:
```diff
@@ -386,6 +474,17 @@ function MapPage() {
     }))
   }, [geoMsgData, msgViewerSliderStartDate, msgViewerSliderEndDate, activeLayers, filter])
 
+  // Effect for handling point source updates moove-ai-layer
+  useEffect(() => {
+    // if the moove-ai-layer is not active, exit the effect
+    if (!activeLayers.includes('moove-ai-layer')) return
+
+    setMooveAiDataSource((prevDataSource) => ({
+      ...prevDataSource,
+      features: mooveAiData,
+    }))
+  }, [mooveAiData, activeLayers, mooveAiFilter])
```
does the mooveAiDataSource need to exist? Can we not just reference mooveAiData directly from the rendered element?

drewjj replied:
Addressed

jacob6838 commented on sample.env:
```diff
@@ -111,6 +112,11 @@ CSM_TARGET_SMTP_SERVER_PORT=587
 
+# Moove AI feature environment variables
+GOOGLE_ACCESS_KEY_NAME=sample_gcp_service_account.json
+GCP_PROJECT_ID=
```
Can we remove the other GCP_PFOJECT_ID (in the jpo_count_metric section)

drewjj replied:
Addressed

jacob6838 reviewed (changes requested):
Works great! There is a duplicate in the sample.env, but other than that I just had some optional optimization suggestions. This is super cool to play with
<img width="724" alt="image" src="https://github.com/user-attachments/assets/87bd3d7f-d443-4565-8145-49821edf8668" />

jacob6838 commented on webapp/src/pages/Map.tsx:
```diff
@@ -731,6 +845,12 @@ function MapPage() {
       type: 'symbol',
       tag: 'intersection',
     },
+    {
```
This can be a backlog task, but it would be nice to have a legend with colors on the map

drewjj replied:
I went ahead and added this now

jacob6838 committed:
Merge branch 'develop' into moove-ai-support

drewjj committed:
Address some pull request comments

drewjj committed:
Address remaining pr requests

drewjj committed:
Add a legend to the moove ai layer once a query is rendered

jacob6838 committed:
Moving mooveAiPolygon(Point)Source to response of useMemo

jacob6838 reviewed (approved):
Looks good to me!

mwodahl reviewed (commented):
This looks great and worked super well when run locally! I had two suggestions before approving:
- Can we update the webapp README with the ENABLE_MOOVE_AI_FEATURES variable?
- It looks like the V2X message viewer and Moove AI viewer toggle the other when one is selected, however, the Configure RSU view does not. Should this be updated?

drewjj committed:
Add to readme

drewjj committed:
Merge branch 'moove-ai-support' of https://github.com/CDOT-CV/jpo-cvmanager into moove-ai-support

drewjj commented:
> This looks great and worked super well when run locally! I had two suggestions before approving:
> 
> * Can we update the webapp README with the ENABLE_MOOVE_AI_FEATURES variable?
> * It looks like the V2X message viewer and Moove AI viewer toggle the other when one is selected, however, the Configure RSU view does not. Should this be updated?

I have addressed your comment regarding the webapp README.

The behavior was intentional since the menus were conflicting. The RSU configuration isn't so I left it as is.

mwodahl reviewed (approved):
Looks good!
